include/qore/intern/BufferedStreamReader.h

```diff
@@ -90,24 +109,69 @@ class BufferedStreamReader : public StreamReader {
          // Won't overflow because the buffer actually has real capacity of bufCapacity+1.
          buf[bufCount] = '\0';
 
+         //printd(5, "BufferedStreamReader::readLine() buf: '%s' bufCount: " QLLD "\n", buf, bufCount);
+
          qore_size_t eolLen;
-         const char* p = findEolInBuffer(*eolstr, eolLen, rc==0);
+         char this_pmatch;
+         const char* p = findEolInBuffer(*eolstr, eolLen, rc==0, this_pmatch);
          assert(eolLen >= 0);
 
          if (p) { // Found end of line.
             assert(p >= buf);
             qore_size_t dataSize = p - buf;
             line->concat(buf, dataSize + (trim ? 0 : eolLen));
-            // Move remaining data from middle to the front of the buffer.
+            assert(bufCount >= (dataSize + eolLen));
+            // Move remaining data to the front of the buffer.
             bufCount -= dataSize + eolLen;
-            memmove(buf, p + eolLen, bufCount);
+            if (bufCount)
+               memmove(buf, p + eolLen, bufCount);
             return line.release();
          }
          else {
             if (rc > 0) {
-               qore_size_t dataSize = (bufCount >= eolLen) ? bufCount - eolLen : 0;
+               if (eol) {
+                  // we read in new data, so here we need to look for a partial eol match at the end of the buffer
```

Comment:
The old tests for the case when an explicit multi-character eol string spanned a buffer were incorrect - the eol string did not span the buffer.  I have just pushed new tests now that test this case correctly, and the old code is now failing in stream-reader.qtest, while the new code is passing in buffered-stream-reader.qtest.

The small buffer size is not "stupid"; you might say it's unrealistic, but first of all it's not our job to determine what arguments Qore users will use to functions, and in any case Qore needs to work properly and predictably with all potential arguments to all functions/methods etc.  Also the small buffer sizes made it very easy to test the edge cases.

Our BufferedStreamReader class is modeled after Java's  [BufferedReader](https://docs.oracle.com/javase/8/docs/api/java/io/BufferedReader.html) which also has a constructor that takes a buffer size, so the buffer size argument is sensible in my opinion.

Please fix the bug in the unbuffered reader as well before merging.